pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/SchemaChangeHandler.java

```diff
@@ -0,0 +1,341 @@
+  }
+
+  SchemaChangeHandler(File indexDir, SegmentMetadataImpl metadata,
+      IndexLoadingConfigMetadata indexConfig, Schema schema) {
+    Preconditions.checkNotNull(indexDir);
+    Preconditions.checkNotNull(metadata);
+    Preconditions.checkState(indexDir.exists(), "Segment directory: {} does not exist", indexDir);
+    Preconditions.checkState(indexDir.isDirectory(), "Segment path: {} is not a directory",
+        indexDir);
+
+    this.indexDir = indexDir;
+    this.metadata = metadata;
+    this.indexConfig = Optional.fromNullable(indexConfig);
+    this.schema = schema;
+    // always use mmap. That's safest and performs well without impact from
+    // -Xmx params
+    // This is not final load of the segment
+    segmentDirectory = SegmentDirectory.createFromLocalFS(indexDir, metadata, ReadMode.mmap);
+  }
+
+  @SuppressWarnings("unchecked")
+  public SegmentMetadataImpl process() throws Exception {
+    if (schema == null) {
+      return metadata;
+    }
+    // compute the action needed for each column
+    Map<String, Action> schemaUpdateActionMap = computeSchemaUpdateAction();
+    //return if schema and segment are in sync
+    if (schemaUpdateActionMap.isEmpty()) {
+      return metadata;
+    }
+    
+    // perform the necessary step for each column
+
+    File tmpIndexDir = new File(indexDir + "/tmp");
+    FileUtils.deleteDirectory(tmpIndexDir);
+    tmpIndexDir.mkdirs();
+    PropertiesConfiguration newMetadataProperties =
+        metadata.getSegmentMetadataPropertiesConfiguration();
+    // create the dictionary and forward index in tmpIndexDir,
+    for (String column : schemaUpdateActionMap.keySet()) {
+      // this method updates the metadataProperties
+      createColumnWithDefaultValue(column, newMetadataProperties, tmpIndexDir);
+    }
+    // update the actual segment
+    Writer segmentWriter = segmentDirectory.createWriter();
+    List<String> newDimensionColumns = new ArrayList<>();
+    List<String> newMetricColumns = new ArrayList<>();
+    for (String column : schemaUpdateActionMap.keySet()) {
+      Action action = schemaUpdateActionMap.get(column);
+      if (action == Action.ADD_DIMENSION) {
+        newDimensionColumns.add(column);
+      } else if (action == Action.ADD_METRIC) {
+        newMetricColumns.add(column);
+      }
+      if (action == Action.UPDATE_DIMENSION_DEFAULT_VALUE
+          || action == Action.UPDATE_METRIC_DEFAULT_VALUE) {
+        if (segmentWriter.isIndexRemovalSupported()) {
+          segmentWriter.removeIndex(column, ColumnIndexType.DICTIONARY);
+          segmentWriter.removeIndex(column, ColumnIndexType.FORWARD_INDEX);
+        } else {
+          LOGGER.warn("Updating default value is not supported in {} segment version",
+              metadata.getVersion());
+        }
+      }
+      // add dictionary
+      String dictionaryFileName = metadata.getDictionaryFileName(column, metadata.getVersion());
```

Comment:
Existence of dictionary/forward index indicates
1. previously failure before updating metadata
2. bad schema (mismatching datatype ? )
3. bad segment metadata
4. bad code

So, relevant case to handle is only 1. We can skip the creation and move forward.